Add typed metavariables for C

emjin opened the pull request:
Closes #1344 

Made changes in pfff (returntocorp/pfff#386) and added test cases for using typed metavariables in equality comparison.

Test plan: make test, metavar_typed.c should have two matches

emjin committed:
Add typed metavariables for C

Closes #2388

Made changes in pfff (returntocorp/pfff#386) and added test cases for using typed metavariables in equality comparison.

Test plan: make test, metavar_typed.c should have two matches

emjin committed:
Leave literal string match as todo

emjin committed:
Merged in develop

emjin committed:
Match char pointers with string literals

mschwager commented:
w00t! I believe this closes https://github.com/returntocorp/semgrep/issues/1344, not https://github.com/returntocorp/semgrep/issues/2388. Regardless, we should close https://github.com/returntocorp/semgrep/issues/2388 and create individual issues for typed metavariables in each requested language.

emjin commented:
> w00t! I believe this closes #1344, not #2388. Regardless, we should close #2388 and create individual issues for typed metavariables in each requested language.

Oops my bad! Edited. I also closed the other issue and added ones for Python and Javascript!

emjin committed:
Treat NULL as any pointer type

Let any pointer type match with NULL in Generic_vs_generic. Added a test for NULL (this test also covers arguments and generic types).

Test plan: make test

emjin committed:
Understand string literals as char pointers

When TypedMetavar matching encounters a metavariable pointer type and a literal string, it should associate the pointer with char pointer

Test plan: make test with updated metavar_typed_generic

Note: Currently all the literals are thrown together---this may be fine, but it's a little ugly. Probably the better thing to do would be to pass in the language and case on the language to say what type to make the literal

aryx commented on semgrep-core/tests/c/metavar_typed_generic.c:
```diff
@@ -0,0 +1,30 @@
+    if ("lit1" == "lit2") return 4;
+
+    //ERROR: match
+    if (NULL == NULL) return 5;
```
I'm curious what $T is in this case.
